### ISSUE TITLE
Cant change the page template with PHP 8

### ISSUE THREAD

**Issue by cb2004** [processwire/processwire-issues#1363]
I have been scratching my head with this one for a while, on some installs I could change the page template no problem, on others I was having to go into the database as the other template options were not available to me in the dropdown. It turns out this could be a problem with PHP 8:

https://processwire.com/talk/topic/25400-cant-change-page-template-after-use-them/

**Comment by ryancramerdesign** [comment]
@cb2004 I am unable to get PHP8 to install successfully on my current development computer, so it may be awhile before I can track this one down. Though if you or anyone else has any idea on the source of it and can point me in the direction, I can investigate further and attempt to workaround it.

**Comment by adrianbj** [comment]
I am on PHP8 and can't seem to replicate this issue. That forum thread suggests something about changing the default language. I tried that and still couldn't reproduce it, but there are so many modules in the ML system that I wonder if it's some combination I don't have?

**Comment by cb2004** [comment]
I get it on all my PHP 8 sites and none of those have multi language. Let me know the best way to help track it down and I can find some time for you no problem.

**Comment by matjazpotocnik** [comment]
@cb2004 I am on PHP8 (windows) and also can't replicate this issue. Can you think of anything else that could cause this behaviour?

**Comment by sekru1** [comment]
I also can't enable PHP8 on my server because then the template can't be changed. This is very annoying.

I was able to narrow down the problem a bit:
There are still templates to choose from that are not currently used by any page.
Maybe this helps you, Ryan?

**Comment by ryancramerdesign** [comment]
@cb2004 @sekru1 I've been around here long enough to know that if adrianbj and matjazpotocnik can't duplicate the issue in PHP8 then I've got no hope to duplicate it myself. :) But I have had a look through the code to see if I can spot anything that might potentially affect behavior in PHP8. The code for this is all pretty simple and standard, and I haven't been able to spot any potential issues yet. But if you'd like to have a look at it, it's all in a method called `ProcessPageEdit::getAllowedTemplates()` and can be viewed here: 

https://github.com/processwire/processwire/blob/dev/wire/modules/Process/ProcessPageEdit/ProcessPageEdit.module#L2660 

Please let me know if you can spot anything? Or if you have a live server running PHP8 and exhibiting the issue,  I could always attempt to debug it from there too.

**Comment by adrianbj** [comment]
@ryancramerdesign - I think I have it figured out. I run all my sites in advanced mode so didn't notice this initially, but it appears in normal mode. The issue is this line:
https://github.com/processwire/processwire/blob/896100100187ad25f9e49dad0c047b01909013f3/wire/modules/Process/ProcessPageEdit/ProcessPageEdit.module#L2709

PHP8 treats `''` differently to `0` when doing comparison operators like `< 0` and I see that `$template->noParents` equals `''`.

https://www.php.net/manual/en/migration80.incompatible.php

Make sense?

**Comment by adrianbj** [comment]
PS - this more strict logic has broken some of my own code as well, but as yet I haven't noticed any other core PW issues, but it seems like there might be some others.

**Comment by ryancramerdesign** [comment]
@adrianbj Thanks, I'll have to have a closer look at it. So the suspect line is this one:
~~~~~
if(!$superAdvanced && $template->noParents < 0 && $template->getNumPages() > 0) {
~~~~~
I think it's definitely strange, but probably okay if it treats `''` differently from `0`. Though it makes me wonder what the point of strict comparison operators is anymore. But in this case it seems like it would have to be treating a blank string as having a negative integer value? I think this issue will definitely pop up in more places (after all, everything returned by MySQL is a string), but I'm glad to hear you've not seen it cause issues elsewhere yet though.  So I'm thinking maybe we change the line to this?
~~~~~
if(!$superAdvanced && ((int) $template->noParents) < 0 && $template->getNumPages() > 0) {
~~~~~
Or will PHP8 typecast the blank string to some negative integer value? If so, maybe we'd need this:
~~~~~
$noParents = $template->noParents === '' ? 0 : intval($template->noParents); 
if(!$superAdvanced && $noParents < 0 && $template->getNumPages() > 0) {
~~~~~
I guess I'm a little afraid to change anything here until I better understand it, so going to do some more reading. Thanks.

**Comment by adrianbj** [comment]
> Or will PHP8 typecast the blank string to some negative integer value?

It casts it to `0`

**Comment by ryancramerdesign** [comment]
That's good. Strange how it compares with a negative number in an if()
statement though. There's got to be more to it, I'll keep reading.

**Comment by teppokoivula** [comment]
It's explained here: http://docs.php.net/manual/en/language.operators.comparison.php.

> **Warning** Prior to PHP 8.0.0, if a string is compared to a number or a numeric string then the string was converted to a number before performing the comparison. This can lead to surprising results as can be seen with the following example: [...]

The key is that strings are no longer converted to numbers in this type of comparison, which means that `''` remains `''` *or* is converted to `null` (not sure if that's specifically declared anywhere), both of which are technically smaller than `0` (and smaller than even `PHP_INT_MIN` for that matter) :)

**Comment by sekru1** [comment]
Thank you all.
I can confirm that this change by Ryan solves the problem:
`if(!$superAdvanced && ((int) $template->noParents) < 0 && $template->getNumPages() > 0) {`

I hope it will find its way into the next release.

**Comment by ryancramerdesign** [comment]
Thanks guys, I've pushed that update to the dev branch. 

> The key is that strings are no longer converted to numbers in this type of comparison, which means that '' remains '' or is converted to null (not sure if that's specifically declared anywhere), both of which are technically smaller than 0 (and smaller than even PHP_INT_MIN for that matter) :)

Thanks for the additional info. Admittedly, I find this change in PHP to be really strange and confusing. Not even JS has that behavior. As far as comparison purposes go, the effect is the same as converting them to a negative number. I'm thinking side effects of that are likely to pop up all over, so will have keep an eye out.

**Comment by cb2004** [comment]
Nice work everyone

**Comment by adrianbj** [comment]
@ryancramerdesign - I am curious why the `noChildren` and `noParents` settings return blank strings when they are not set. Shouldn't they really return `0` ?

That said, I don't really understand why `(int) $template->noParents < 0` is working as expected. If I do that check on a template, the integer of `$template->noParents` returns `0` and the conditional `<` against `0` returns `false`.

![image](https://user-images.githubusercontent.com/871211/127516340-76156923-f7ce-48d8-a955-827f5dabf610.png)

**Comment by matjazpotocnik** [comment]
![image](https://user-images.githubusercontent.com/7663992/127750530-2c894f18-5365-4176-a797-fc8bddf2a35f.png)

**Comment by cb2004** [comment]
This now works as expected for me but should I reopen?

**Comment by teppokoivula** [comment]
>  I am curious why the noChildren and noParents settings return blank strings when they are not set. Shouldn't they really return 0 ?

This is a good point, especially since the PHP doc comments state that the value is an integer. For consistency it would be good if either the values were always integer (probably preferable) or alternatively the docs were updated to reflect actual type (string|int) with an explanation why/when the value is a string.

This is, though, mostly unrelated to the original issue here — perhaps open a new issue for this?

> That said, I don't really understand why (int) $template->noParents < 0 is working as expected. If I do that check on a template, the integer of $template->noParents returns 0 and the conditional < against 0 returns false.

This is covered by the docs.

When you cast noParents to integer, you're essentially duplicating the behaviour from PHP < 8. This is what used to happen behind the scenes automatically, but since it could result in really obscure issues, they're no longer doing it (but one can, of course, reproduce the same effect if/when it makes sense).

If you don't cast to an integer the empty value is used in the comparison, and this'll always be smaller than any possible integer value.

**Comment by adrianbj** [comment]
@teppokoivula - maybe I am just being too literal, but if you cast an empty string as an integer it returns `0` so if you compare `0 <  0` that returns `false` which is what I (and @matjazpotocnik ) showed above. This is why I don't get why Ryan's recent changes actually work, because I would have thought it should be `(int) $template->noParents) < 1` or `(int) $template->noParents) === 0`

But it sounds like it is working, so obviously I am missing something and I don't have the energy to dive in deeper, so feel free to ignore me on this :)

**Comment by teppokoivula** [comment]
> because I would have thought it should be `(int) $template->noParents) < 1` or `(int) $template->noParents) === 0`

The fixed version checks if `((int) $template->noParents) < 0`. In case noParents is an empty string, this is essentially the same as `(0) < 0` — which is functionally identical to how the old version (`$template->noParents < 0`) was interpreted in PHP < 8 (before PHP 8 `'' < 0` was internally translated to `0 < 0`).

In other words this change just duplicates the PHP < 8 comparison logic for PHP 8. Valid values for `noParents` are `-1`, `0` (or empty string), and `1`, out of which this will only evaluate to `true` in case the value is `-1` :)

**Comment by adrianbj** [comment]
Thanks @teppokoivula - sorry, I was not paying proper attention to what happened in that conditional if it was true (so I was getting things backwards) and I also forgot about the `-1` option for when it is set to "One" new page.

Sorry to take up your time.

**Comment by klor** [comment]
Should anyone want to patch their installation manually until a fix for Master (3.0.165) has been released, then it can be done in ProcessPageEdit.module by changing L2717 and L2722. In both lines, change "$template->noParents" ... to "**(int)** $template->noParents".
